### PR TITLE
Add a tutorial on how to add a plugin sidebar

### PR DESCRIPTION
**Pull request by oandregal**
This PR aims to add a new tutorial: **How to add a plugin sidebar**.

[View docs live](https://github.com/WordPress/gutenberg/blob/add/tutorial-sidebar/docs/designers-developers/developers/tutorials/sidebar-tutorial/plugin-sidebar-0.md).

Note that the images have been updated to point to the production URLs, such as https://raw.githubusercontent.com/WordPress/gutenberg/master/docs/designers-developers/assets/image.png so they won't be shown inline in the live docs above because they haven't been deployed yet.

### CODE REVIEW

**Comment by chrisvanpatten** [comment]
Love it — let me know when it's ready for review!

File: docs/designers-developers/developers/tutorials/block-tutorial/plugin-sidebar.md [hunk 1]
```diff
@@ -124,6 +124,7 @@ add_action( 'enqueue_block_assets', 'sidebar_plugin_style_enqueue' );
 
 ## Step 3: an input that updates its content
 
+```js
```
Comment (chrisvanpatten) [inline docs/designers-developers/developers/tutorials/block-tutorial/plugin-sidebar.md]: Might be good to check that the markdown parser on dot-org supports language-hinted code fences.

Comment (nosolosw) [reply docs/designers-developers/developers/tutorials/block-tutorial/plugin-sidebar.md]: It looks like it does. [This tutorial](https://github.com/WordPress/gutenberg/blob/master/docs/designers-developers/developers/tutorials/block-tutorial/writing-your-first-block-type.md) uses the php code tag and is rendered properly in the [handbook](https://wordpress.org/gutenberg/handbook/designers-developers/developers/tutorials/block-tutorial/writing-your-first-block-type/).

**Comment by nosolosw** [comment]
Just a note that we're considering merging this with the Metabox tutorial that it's [in the works](https://github.com/WordPress/gutenberg/pull/12834) as well.

**Comment by nosolosw** [comment]
This is ready to review! I still think this will make a good addition within the Metaboxes tutorial by @mkaz but wanted to hear other opinions first, given how long it ended up being.

**Comment by nosolosw** [comment]
We seem to want this [as a standalone tutorial](https://wordpress.slack.com/archives/C02RP4WU5/p1545224966816000) due to its length. I've updated the toc, manifest, moved images to the assets folder, and updated links accordingly. Until reviews come in, that's it on my end!

**Review by mkaz — CHANGES_REQUESTED** [review]
Great work Andres! It is looking really good and thorough, this will be an excellent resource.

First up, it is recommended to write in the second person, so instead of "we" it would just be "you".
From: https://make.wordpress.org/docs/handbook/documentation-team-handbook/handbooks-style-and-formatting-guide/

Here are my notes going through each step:

### Step 0:

Move the plugin reference to the top:

This tutorial starts with you having an existing plugin setup and ready to add PHP and JavaScript code. Please refer to Getting started with JavaScript tutorial for an introduction to WordPress plugins and how to use JavaScript to extend the block editor.

### Step 1:

It is good to be clear on where to put the code, at least the first time or two. Add the following code to a JavaScript file called `sidebar-plugin.js`. The PHP example is more clear since it starts with `<?php` and doesn't matter what it is named.


Is this syntax necessary? 
```
( function( wp ) {
...
}( window.wp );
```

It looks a little foreign to someone who might be new. Could we use `wp.domReady( fuction() { })` which might seem more familiar to a developer coming from jQuery. I don't feel strongly on this one, just first time I saw it I was surprised.


### Step 2:

Be clear what they should do with the code. Replace the existing code in `sidebar-plugin.js` with this updated code.

Feedback: I got cute and thought I knew what I was doing and only tried copy and pasting the CSS enqueue part, but it didn't work since I missed the register script part. 

Not sure if that is something that needs changing, or not. 


### Step 3:

I like the use of browser console to check the field is registered.

### Step 4:

Minor reword:

`withSelect` is used to pass data to other components, and update when the original data changes.

In the bullet list at bottom, I would move the dependency up so it doesn't get missed. The others explain the code, but that is a step the user must do.

### Step 5:

The explanation for the changes are listed within the comments of the code which is fine. however in step 4, most of the explanation was in the bullet list below.  

I don't have a preference, but I would make the two consistent and then if it is in the code as comments, call out in the tutorial text to read the comments around the new code.


### Step 6:

I would reword the heading to something like:

Your JavaScript code now works as expected, here are a few ways to refactor the code to simplfiy and make it easier to change in the future.

**Comment by nosolosw** [comment]
Great feedback, Marcus.

> Is this syntax necessary?
> 
> ( function( wp ) {
> ...
> }( window.wp );

My take is that plugins should avoid polluting the global scope. I've used IIFEs assuming it was the norm, but I don't know whether there is a _WordPress way_ or plugins just use IIFEs. I'll look that up and will update accordingly.

File: docs/designers-developers/developers/tutorials/sidebar-tutorial/plugin-sidebar-1-up-and-running.md [hunk 2]
```diff
@@ -0,0 +1,46 @@
+# Get a sidebar up and running
+
+First, we need to tell the editor that we are registering a new plugin that will have its own sidebar. We can do so by using the [`wp.plugins.registerPlugin`](https://wordpress.org/gutenberg/handbook/designers-developers/developers/packages/packages-plugins/) and [`wp.editPost.PluginSidebar`](https://wordpress.org/gutenberg/handbook/designers-developers/developers/packages/packages-edit-post/#pluginsidebar) utilities.
+
+```js
+( function( wp ) {
+	wp.plugins.registerPlugin( 'my-plugin-sidebar', {
+		render: function(){
+			return wp.editPost.PluginSidebar( {
```
Comment (gziolo) [inline docs/designers-developers/developers/tutorials/sidebar-tutorial/plugin-sidebar-1-up-and-running.md]: I think you should be using `wp.element.createElement` in examples. At least it is how JSX would work:
```
return wp.element.createElement(
    wp.editPost.PluginSidebar,
    props,
    'Meta field'
);
```

File: docs/designers-developers/developers/tutorials/sidebar-tutorial/plugin-sidebar-2-styles-and-controls.md [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+After the sidebar is up and running, we can focus on filling it up with the necessary components and basic styling. To visualize and edit the meta field value we'll use a input component. The [`wp.components` package](https://wordpress.org/gutenberg/handbook/designers-developers/developers/packages/packages-components/) contains many components available for us to reuse, and, specifically, the [`TextControl`](https://wordpress.org/gutenberg/handbook/designers-developers/developers/components/text-control/) let us create an input field, so that's what we'll use:
+
+```js
+( function( wp ) {
+	var el = wp.element.createElement;
+	var Text = wp.components.TextControl;
+
+	wp.plugins.registerPlugin( 'my-plugin-sidebar', {
+		render: function(){
+			return wp.editPost.PluginSidebar( {
```
Comment (gziolo) [inline docs/designers-developers/developers/tutorials/sidebar-tutorial/plugin-sidebar-2-styles-and-controls.md]: The same thing with `el`

File: docs/designers-developers/developers/tutorials/sidebar-tutorial/plugin-sidebar-2-styles-and-controls.md [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+
+function sidebar_plugin_script_enqueue() {
+	wp_enqueue_script( 'sidebar-plugin-js' );
+}
+add_action( 'enqueue_block_editor_assets', 'sidebar_plugin_script_enqueue' );
+
+function sidebar_plugin_style_enqueue() {
+	wp_enqueue_style( 'sidebar-plugin-css' );
+}
+add_action( 'enqueue_block_assets', 'sidebar_plugin_style_enqueue' );
```
Comment (gziolo) [inline docs/designers-developers/developers/tutorials/sidebar-tutorial/plugin-sidebar-2-styles-and-controls.md]: This doesn't seem right to use `enqueue_block_assets` here as it isn't tied to any block. Not sure what is the correct way though.

File: docs/designers-developers/developers/tutorials/sidebar-tutorial/plugin-sidebar-4-initialize-input.md [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+			value: 'Initial value',
+			onChange: function( content ) {
+				console.log( 'content changed to ', content );
+			},
+		});
+	}
+
+	wp.plugins.registerPlugin( 'my-plugin-sidebar', {
+		render: function(){
+			return wp.editPost.PluginSidebar( {
```
Comment (gziolo) [inline docs/designers-developers/developers/tutorials/sidebar-tutorial/plugin-sidebar-4-initialize-input.md]: Again, this should go through `el` as everything else.

File: docs/designers-developers/developers/tutorials/sidebar-tutorial/plugin-sidebar-4-initialize-input.md [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+`withSelect` is useful to pass some data to other components, and update them when the original data changes. Let's use it in our plugin code:
+
+```js
+( function( wp ) {
+	var el = wp.element.createElement;
+	var Text = wp.components.TextControl;
+	var withSelect = wp.data.withSelect;
+
+	// Function that takes `select` as input and returns some data.
+	var selectToData = function( select ) {
```
Comment (gziolo) [inline docs/designers-developers/developers/tutorials/sidebar-tutorial/plugin-sidebar-4-initialize-input.md]: This function is called `mapSelectToProps` in `withSelect` definition.

File: docs/designers-developers/developers/tutorials/sidebar-tutorial/plugin-sidebar-4-initialize-input.md [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+				console.log( 'content has changed to ', content );
+			},
+		} );
+	}
+
+	var MetaBlockFieldWithData = withSelect( selectToData )( MetaBlockField );
+
+	wp.plugins.registerPlugin( 'my-plugin-sidebar', {
+		render: function(){
+			return wp.editPost.PluginSidebar( {
```
Comment (gziolo) [inline docs/designers-developers/developers/tutorials/sidebar-tutorial/plugin-sidebar-4-initialize-input.md]: `el` again :)

**Review by gziolo — COMMENTED** [review]
Nice works on this one, I didn't read through all of that. However, I hope to do it once it's ready :)

I love the fact that those tutorials are all ES5 💯

File: docs/designers-developers/developers/tutorials/sidebar-tutorial/plugin-sidebar-4-initialize-input.md (hunk 6 above)
Comment (nosolosw) [reply docs/designers-developers/developers/tutorials/sidebar-tutorial/plugin-sidebar-4-initialize-input.md]: 77abb82 and 174fccf

File: docs/designers-developers/developers/tutorials/sidebar-tutorial/plugin-sidebar-2-styles-and-controls.md (hunk 4 above)
Comment (nosolosw) [reply docs/designers-developers/developers/tutorials/sidebar-tutorial/plugin-sidebar-2-styles-and-controls.md]: Had the same thought, but don't know of any other way to enqueue styles both for the editor and the front-end.

File: docs/designers-developers/developers/tutorials/sidebar-tutorial/plugin-sidebar-1-up-and-running.md (hunk 2 above)
Comment (nosolosw) [reply docs/designers-developers/developers/tutorials/sidebar-tutorial/plugin-sidebar-1-up-and-running.md]: f53b7e2

File: docs/designers-developers/developers/tutorials/sidebar-tutorial/plugin-sidebar-5-update-meta.md [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+			},
+		} );
+	}
+
+	var MetaBlockFieldWithData = withSelect( mapSelectToProps )( MetaBlockField );
+	var MetaBlockFieldWithDataAndActions = withDispatch( mapDispatchToProps )( MetaBlockFieldWithData );
+
+	registerPlugin( 'my-plugin-sidebar', {
+		render: function() {
+			return el( PluginSidebar),
```
Comment (gziolo) [inline docs/designers-developers/developers/tutorials/sidebar-tutorial/plugin-sidebar-5-update-meta.md]: Closing `)` should be removed :)

File: docs/designers-developers/developers/tutorials/sidebar-tutorial/plugin-sidebar-6-finishing-touches.md [hunk 9]
```diff
@@ -0,0 +1,198 @@
...
+						{ meta: { sidebar_plugin_meta_block_field: value } }
+					);
+				}
+			}
+		}
+	)( MetaBlockFieldWithData );
+
+	registerPlugin( 'my-plugin-sidebar', {
+		render: function() {
+			return el( PluginSidebar),
```
Comment (gziolo) [inline docs/designers-developers/developers/tutorials/sidebar-tutorial/plugin-sidebar-6-finishing-touches.md]: Closing `)` should be removed here as well

File: docs/designers-developers/developers/tutorials/sidebar-tutorial/plugin-sidebar-5-update-meta.md (hunk 8 above)
Comment (nosolosw) [reply docs/designers-developers/developers/tutorials/sidebar-tutorial/plugin-sidebar-5-update-meta.md]: Good catch! 😅

**Comment by nosolosw** [comment]
I've created https://github.com/WordPress/gutenberg/pull/13041 following up a convo with Marcus about how to explain IIFEs to new developers.

**Comment by chrisvanpatten** [comment]
On my list to review today

File: docs/designers-developers/developers/tutorials/sidebar-tutorial/plugin-sidebar-2-styles-and-controls.md (hunk 4 above)
Comment (chrisvanpatten) [reply docs/designers-developers/developers/tutorials/sidebar-tutorial/plugin-sidebar-2-styles-and-controls.md]: @nosolosw You can call the enqueue function multiple times, e.g. 

```php
add_action( 'admin_enqueue_scripts', 'sidebar_plugin_style_enqueue' );
add_action( 'wp_enqueue_scripts', 'sidebar_plugin_style_enqueue' );
```

Comment (nosolosw) [reply docs/designers-developers/developers/tutorials/sidebar-tutorial/plugin-sidebar-2-styles-and-controls.md]: Not convinced about using `admin_enqueue_scripts` as it'll require an extra bit of complexity not useful for the example (checking what admin page is loading and back off if it isn't the editor). Unless there is a strong preference about recommending that (or another approach), I'd leave it as it is.

File: docs/designers-developers/developers/tutorials/sidebar-tutorial/plugin-sidebar-0.md [hunk 10]
```diff
@@ -0,0 +1,12 @@
+# Creating a sidebar for your plugin
+
+This tutorial starts with you having an existing plugin setup and ready to add PHP and JavaScript code. Please, refer to [Getting started with JavaScript](https://wordpress.org/gutenberg/handbook/designers-developers/developers/tutorials/javascript/) tutorial for an introduction to WordPress plugins and how to use JavaScript to extend the block editor.
+
+In the next sections, you're going to create a custom sidebar for a plugin that contains a text control so the user cane update a value that is stored in the `post_meta` table.
```
Comment (nosolosw) [inline docs/designers-developers/developers/tutorials/sidebar-tutorial/plugin-sidebar-0.md]:
```suggestion
 In the next sections, you're going to create a custom sidebar for a plugin that contains a text control so the user can update a value that is stored in the `post_meta` table.
```

**Comment by nosolosw** [comment]
Given feedback has been addressed, does this look ready? _Just a ping so this shows up in everyone's 2019 radar :)_

**Review by mkaz — APPROVED** [review]
Sorry about so many commits, I was using Github's edit in place function since I thought there was only one or two changes, but needed a couple more so it became a bit more verbose

**Review by gziolo — COMMENTED** [review]
Content wise it looks great. Great job on the tutorial 💯 

I left some comments related to urls which is something we need to finally solve and apply consistently.

File: docs/designers-developers/developers/tutorials/sidebar-tutorial/plugin-sidebar-2-styles-and-controls.md [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+To visualize and edit the meta field value you'll use an input component. The `@wordpress/components` package contains many components available for you to reuse, and, specifically, the [TextControl](https://wordpress.org/gutenberg/handbook/designers-developers/developers/components/text-control/) is aimed at creating an input field:
+
+```js
+( function( wp ) {
+	var registerPlugin = wp.plugins.registerPlugin;
+	var PluginSidebar = wp.editPost.PluginSidebar;
+	var el = wp.element.createElement;
+	var Text = wp.components.TextControl;
+
+	registerPlugin( 'my-plugin-sidebar', {
```
[the remark this answers is not in the document]
Comment (nosolosw) [reply docs/designers-developers/developers/tutorials/sidebar-tutorial/plugin-sidebar-2-styles-and-controls.md]: Updated.

File: docs/designers-developers/developers/tutorials/sidebar-tutorial/plugin-sidebar-1-up-and-running.md [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+
+function sidebar_plugin_script_enqueue() {
+	wp_enqueue_script( 'sidebar-plugin-js' );
+}
+add_action( 'enqueue_block_editor_assets', 'sidebar_plugin_script_enqueue' );
+```
+
+After installing and activating this plugin, there is a new icon resembling a tack in the top-right of the editor. Upon clicking it, the plugin's sidebar will be opened:
+
+![Sidebar Up and Running](https://raw.githubusercontent.com/WordPress/gutenberg/master/docs/designers-developers/assets/sidebar-up-and-running.png)
```
[the remark this answers is not in the document]
Comment (nosolosw) [reply docs/designers-developers/developers/tutorials/sidebar-tutorial/plugin-sidebar-1-up-and-running.md]: This is because it's pointing to the URL they'll have when publishing (see comment on description). So far my understanding goes, this is the way we're managing images in the handbook. Happy to change if there is a better way.

File: docs/designers-developers/developers/tutorials/sidebar-tutorial/plugin-sidebar-0.md [hunk 13]
```diff
@@ -0,0 +1,12 @@
+# Creating a sidebar for your plugin
+
+This tutorial starts with you having an existing plugin setup and ready to add PHP and JavaScript code. Please, refer to [Getting started with JavaScript](https://wordpress.org/gutenberg/handbook/designers-developers/developers/tutorials/javascript/) tutorial for an introduction to WordPress plugins and how to use JavaScript to extend the block editor.
+
+ In the next sections, you're going to create a custom sidebar for a plugin that contains a text control so the user can update a value that is stored in the `post_meta` table.
+
+1. [Get a sidebar up and running](./plugin-sidebar-1-up-and-running.md)
```
[the remark this answers is not in the document]
Comment (nosolosw) [reply docs/designers-developers/developers/tutorials/sidebar-tutorial/plugin-sidebar-0.md]: Fixed.

File: docs/designers-developers/developers/tutorials/sidebar-tutorial/plugin-sidebar-2-styles-and-controls.md [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+
+function sidebar_plugin_style_enqueue() {
+	wp_enqueue_style( 'sidebar-plugin-css' );
+}
+add_action( 'enqueue_block_assets', 'sidebar_plugin_style_enqueue' );
+```
+
+Reload the editor and open the sidebar:
+
+![Sidebar with style and controls](https://raw.githubusercontent.com/WordPress/gutenberg/master/docs/designers-developers/assets/sidebar-style-and-controls.png)
```
[the remark this answers is not in the document]
Comment (nosolosw) [reply docs/designers-developers/developers/tutorials/sidebar-tutorial/plugin-sidebar-2-styles-and-controls.md]: Same as above

**Comment by gziolo** [comment]
There are more link which use hardcoded handbook URL:
`https://wordpress.org/gutenberg/handbook/...`

Can we fix all of them and proceed with merge?

**Comment by nosolosw** [comment]
Do we need to update `https://wordpress.org/gutenberg/handbook/` to `../../docs`? Does it work? I see the URL being used in other tutorials and I thought it'd be safer (if the URL changed, WordPress will take care of redirections). I'm not familiar with how these links work. I can change them to anything, just want to make sure they work once the tutorial is live.

**Comment by mkaz** [comment]
@nosolosw I've been using the `../../docs` syntax going back to top level which allows the documentation to be browsed both on Github and the generated site. I think absolute refs can be used instead of relative, but that will only work on generated site and not Github preview.

**Comment by nosolosw** [comment]
I've updated the files to use relative paths instead of full URLs to the handbook. I'm not familiar with how the handbook is built, so here are some places where GitHub links are working and I wonder if relative paths would be OK in production:

* references to a subpath under packages (which doesn't exist as a subfolder under the handbook hierarchy), such as https://wordpress.org/gutenberg/handbook/designers-developers/developers/packages/packages-plugins/ [Example](https://github.com/WordPress/gutenberg/blob/add/tutorial-sidebar/docs/designers-developers/developers/tutorials/sidebar-tutorial/plugin-sidebar-1-up-and-running.md).
* references to components https://wordpress.org/gutenberg/handbook/designers-developers/developers/components/ [Example](https://github.com/WordPress/gutenberg/blob/add/tutorial-sidebar/docs/designers-developers/developers/tutorials/sidebar-tutorial/plugin-sidebar-2-styles-and-controls.md).
* references to data. [Examples](https://github.com/WordPress/gutenberg/blob/add/tutorial-sidebar/docs/designers-developers/developers/tutorials/sidebar-tutorial/plugin-sidebar-4-initialize-input.md).

**Review by gziolo — APPROVED** [review]
I fixed all urls in a way they work with both GitHub and Gutenberg handbook. You can double check patterns used in the README files of the packages.

The easiest way to verify is to use view file link from the `Files changed` tab, example:

https://github.com/WordPress/gutenberg/blob/7ebecd592361a191bb8390e99dd1f2b2a7e632a9/docs/designers-developers/developers/tutorials/sidebar-tutorial/plugin-sidebar-0.md

**Comment by nosolosw** [comment]
Thanks, Grzegorz!
